listener.go

```diff
@@ -208,20 +207,16 @@ func (l *Listener) Listen(ctx context.Context, handle Handle, topicName string,
 	// Create a handle class that has that function
 	listenerHandle := subReceiver.Listen(ctx, servicebus.HandlerFunc(
 		func(ctx context.Context, message *servicebus.Message) error {
-			if val, ok := message.UserProperties["type"]; ok {
-				err := handle(ctx, string(message.Data), val.(string))
-				if err != nil {
-					err = message.Abandon(ctx)
-					return err
+			originalHandler := handler
+			for !handle.IsDone(handler) {
+				handler = handler.Do(ctx, originalHandler, message)
+				// handle nil as a Completion!
```

Comment:
is this good? do we want to try and force them to always say compelte or abandon or retry

Reply:
Without a doubt yes! Force the user to make write good code, with well defined behavior.
Note that :
- Abandon = leave the message on the queue -> message will be reprocessed (unless max retry is reached)
- Complete= I'm done processing, delete the message.

there is no other choice (not considering Defer or special functionality not exposed here)